### PR TITLE
Handle changes to `MutableSettings` and `ExporterSettings` without rebuilding

### PR DESCRIPTION
**Pull request by andrewlock**
## Summary of changes

- This is the big one
- Update services to dynamically update when mutable settings or exporter settings change
- Stop rebuilding everything when there's manual/remote configuration

## Reason for change

This is the "endpoint" that we've been heading for - services only being disposed/rebuilt at the end of the app, and otherwise only rebuilding the _necessary_ parts. For example - we don't need to tear down all the API factories when a customer changes a global tag via remote config; they only need to change if the `ExporterSettings` change.

The hope is that overall this reduces the overhead of using configuration in code and/or remote configuration, while also reducing the number of issues due to managing disposal of services.

## Implementation details

Overall, this PR is kind of a pain. Moving from the "rebuild everything" to "reconfigure each service" couldn't be done piecemeal, so this is the one-shot PR. What's more, different services need different patterns (though we can probably consolidate some of them, this has taken a _lot_ of work and I likely changed patterns unnecessarily in some places).

In general, there's a couple of patterns:

- CI Vis doesn't let you change settings at runtime, so it _never_ needs to respond to changes. It always just uses the "initial" settings
- Debugger _today_ doesn't respond to changes at runtime (except its own dynamic config), so for now we ignore Debugger too as it's not really a regression. I hope we can fix this soon though.
- I've introduced the concept of `Managed*` versions of some services
  - These services generally "wrap" the existing type, delegating access to the underlying service, and handling settings changes
- Many services only care about a sub-set of mutable settings, so they only update if they need to
- Somewhat annoyingly, setting updates occur on a background thread, so we need to be careful about thread safety. Where necessary (most places) I've made sure access to a now-mutable service is done using `Volatile.Read()` (to ensure changes are visible) and are generally cached to a local variable (as the underlying field may be updated in the background).

## Test coverage

In the vast majority of places, this should be covered by existing tests

I plan to add some additional integration tests around reconfiguring and a bunch of manual testing to make sure I'm confident.

## Other details

I strongly recommend reviewing commit-by-commit. They're generally self-contained, and hopefully simple enough to understand one commit at a time.

https://datadoghq.atlassian.net/browse/LANGPLAT-819

Part of a config stack

- https://github.com/DataDog/dd-trace-dotnet/pull/7522
- https://github.com/DataDog/dd-trace-dotnet/pull/7525
- https://github.com/DataDog/dd-trace-dotnet/pull/7530
- https://github.com/DataDog/dd-trace-dotnet/pull/7532
- https://github.com/DataDog/dd-trace-dotnet/pull/7543
- https://github.com/DataDog/dd-trace-dotnet/pull/7544
- https://github.com/DataDog/dd-trace-dotnet/pull/7721
- https://github.com/DataDog/dd-trace-dotnet/pull/7722
- https://github.com/DataDog/dd-trace-dotnet/pull/7695
- https://github.com/DataDog/dd-trace-dotnet/pull/7723
- https://github.com/DataDog/dd-trace-dotnet/pull/7724 👈
- https://github.com/DataDog/dd-trace-dotnet/pull/7796

This isn't the final PR in the stack, as there will be a bunch of cleaning up to do, but it's the final "implementation" PR

### CODE REVIEW

File: tracer/src/Datadog.Trace/DogStatsd/StatsdManager.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    internal static bool HasImpactingChanges(TracerSettings.SettingsManager.SettingChanges changes)
+    {
+        var hasChanges = changes.UpdatedExporter is not null // relying on this to only be non null if _anything_ changed
+                      || (changes.UpdatedMutable is { } updated
+                       && !(
+                               string.Equals(updated.Environment, changes.PreviousMutable.Environment, StringComparison.Ordinal)
+                            && string.Equals(updated.ServiceVersion, changes.PreviousMutable.ServiceVersion, StringComparison.Ordinal)
+                               // The service name comparison isn't _strictly_ correct, because we normalize it further, but this is probably good enough
+                            && string.Equals(updated.DefaultServiceName, changes.PreviousMutable.DefaultServiceName, StringComparison.OrdinalIgnoreCase)
+                            && updated.GlobalTags.SequenceEqual(changes.PreviousMutable.GlobalTags)));
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/DogStatsd/StatsdManager.cs]: That's not a Linq extension, it's the method added in https://github.com/DataDog/dd-trace-dotnet/pull/7722

File: tracer/src/Datadog.Trace/Agent/Api.cs [hunk 2]
```diff
@@ -76,6 +80,12 @@ private enum SendResult
             Failed_DontRetry,
         }
 
+        [MemberNotNull(nameof(_statsd))]
+        public void ToggleTracerHealthMetrics(bool enabled)
+        {
+            _statsd = enabled ? _originalStatsd : NoOpStatsd.Instance;
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/Api.cs]: Thread safety issue: The assignment to `_statsd` should use `Volatile.Write(ref _statsd, enabled ? _originalStatsd : NoOpStatsd.Instance)` or `Interlocked.Exchange` to ensure proper memory visibility across threads. Without a memory barrier, other threads may not see the updated value.

Since this method is called from a settings change callback (which runs on a background thread), and `_statsd` is accessed from other threads, proper synchronization is critical.

*Automated comment by Claude Code*

File: tracer/src/Datadog.Trace/Agent/ManagedApi.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    private Api _api;
+
+    public ManagedApi(
+        TracerSettings.SettingsManager settings,
+        IDogStatsd statsd,
+        Action<Dictionary<string, float>> updateSampleRates,
+        bool partialFlushEnabled)
+    {
+        UpdateApi(settings.InitialExporterSettings, settings.InitialMutableSettings.TracerMetricsEnabled);
+        settings.SubscribeToChanges(changes =>
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/ManagedApi.cs:34]: Memory leak: The `ManagedApi` class subscribes to settings changes but never disposes the subscription. Consider implementing `IDisposable` and disposing the subscription, or document why this is intentional (e.g., if the lifetime matches the application lifetime).

Pattern to follow:
```csharp
private readonly IDisposable _settingsSubscription;

public void Dispose()
{
    _settingsSubscription?.Dispose();
}
```

*Automated comment by Claude Code*

File: tracer/src/Datadog.Trace/LibDatadog/DataPipeline/ManagedTraceExporter.cs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            {
+                var exporter = CreateTraceExporter(
+                    settings,
+                    changes.UpdatedMutable ?? changes.PreviousMutable,
+                    changes.UpdatedExporter ?? changes.PreviousExporter,
+                    updateSampleRates,
+                    telemetrySettings);
+
+                var previous = Interlocked.Exchange(ref _current, exporter);
+                previous?.Dispose();
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/LibDatadog/DataPipeline/ManagedTraceExporter.cs:60]: Potential use-after-dispose: The previous exporter is disposed immediately after exchange on line 55. However, there could be in-flight operations still using the old exporter. Consider either:

1. Deferring disposal with a delay (similar to `StatsdManager.cs:67-72`)
2. Implementing a graceful shutdown mechanism
3. Documenting why immediate disposal is safe here (e.g., if the API methods handle null gracefully)

The pattern in StatsdManager provides a good reference for safer disposal.

*Automated comment by Claude Code*

File: tracer/src/Datadog.Trace/DogStatsd/StatsdManager.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            if (!HasImpactingChanges(c))
+            {
+                return;
+            }
+
+            IDogStatsd previous;
+
+            lock (_lock)
+            {
+                if (Volatile.Read(ref _isDisposed))
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/DogStatsd/StatsdManager.cs]: Potential race condition: The check `Volatile.Read(ref _isDisposed)` on line 50 is inside the lock, but then `_current` is used outside the lock on line 56. If disposal occurs between releasing the lock and calling `StatsdFactory.CreateDogStatsdClient`, the new instance could be created after disposal.

Consider restructuring to either:
1. Keep the entire operation inside the lock
2. Return early before the lock if disposed
3. Re-check disposal after updating `_current` but before exiting the lock

*Automated comment by Claude Code*

File: tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs [hunk 6]
```diff
@@ -27,22 +28,38 @@ internal class DataStreamsManager
     private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<DataStreamsManager>();
     private static readonly AsyncLocal<PathwayContext?> LastConsumePathway = new(); // saves the context on consume checkpointing only
     private readonly ConcurrentDictionary<string, RateLimiter> _schemaRateLimiters = new();
-    private readonly NodeHashBase _nodeHashBase;
+    private readonly IDisposable _updateSubscription;
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs:30]: Memory leak: The `DataStreamsManager` subscribes to settings changes via `_updateSubscription` but never disposes of it. The `DataStreamsManager` class should implement `IDisposable` and dispose the subscription in the `Dispose` method.

```csharp
public void Dispose()
{
    _updateSubscription?.Dispose();
    _writer?.Dispose(); // if applicable
}
```

*Automated comment by Claude Code*

File: tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs [hunk 7]
```diff
@@ -32,16 +32,35 @@ internal class DiscoveryService : IDiscoveryService
...
+        public DiscoveryService(
+            TracerSettings.SettingsManager settings,
+            TimeSpan tcpTimeout,
+            int initialRetryDelayMs,
+            int maxRetryDelayMs,
+            int recheckIntervalMs)
+            : this(CreateApiRequestFactory(settings.InitialExporterSettings, tcpTimeout), initialRetryDelayMs, maxRetryDelayMs, recheckIntervalMs)
+        {
+            // Create as a "managed" service that can update the request factory
+            settings.SubscribeToChanges(changes =>
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs]: Memory leak: The `DiscoveryService` constructor subscribes to settings changes but the subscription is not stored or disposed. Consider storing the subscription as a field and disposing it in `DisposeAsync`.

```csharp
private readonly IDisposable? _settingsSubscription;

public DiscoveryService(TracerSettings.SettingsManager settings, ...)
{
    // ...
    _settingsSubscription = settings.SubscribeToChanges(changes => { ... });
}

public Task DisposeAsync()
{
    _settingsSubscription?.Dispose();
    // ... existing disposal logic
}
```

*Automated comment by Claude Code*

File: tracer/src/Datadog.Trace/Sampling/ManagedTraceSampler.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                var rates = Volatile.Read(ref _defaultSampleRates);
+                if (rates is not null)
+                {
+                    newSampler.SetDefaultSampleRates(rates);
+                }
+
+                Interlocked.Exchange(ref _current, newSampler);
+
+                // There's a small edge case here where the agent rates are being updated at the same time as we're
+                // updating the sampler, so check and just update them again if necessary
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Sampling/ManagedTraceSampler.cs]: Race condition with double-check pattern: Lines 48-52 implement a double-check to handle concurrent updates to `_defaultSampleRates`. However, this pattern has a subtle issue: between line 45 (Exchange) and line 49 (Volatile.Read), another thread could update `_defaultSampleRates` twice, and only the second update would be applied.

While this is unlikely to cause serious issues (since it's just applying the same rates again), consider whether the complexity of the double-check is justified, or if a simpler approach (accepting that very rare concurrent updates might miss one setting) would be clearer.

*Automated comment by Claude Code*

File: tracer/src/Datadog.Trace/Agent/ManagedApi.cs (hunk 3 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/ManagedApi.cs:34]: Yeah, there's a bunch of these - in _general_ the lifetime of all subscribers _will_ match the application lifetime. Will document it for now, but maybe we should consider not even bothering to have "unsubscribe" as an option? 🤔

File: tracer/src/Datadog.Trace/LibDatadog/DataPipeline/ManagedTraceExporter.cs (hunk 4 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/LibDatadog/DataPipeline/ManagedTraceExporter.cs:60]: Yeah, I agonized a bit about this one. Will document my conclusion

File: tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs (hunk 6 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs:30]: There's a `DisposeAsync()` method that's called. That's good enough

File: tracer/src/Datadog.Trace/DogStatsd/StatsdManager.cs (hunk 5 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/DogStatsd/StatsdManager.cs]: > but then _current is used outside the lock on line 56

No it's not? 🤔 In fact, realised we don't need the volatile if we're inside the lock, so removed those calls

File: tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs (hunk 7 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs]: It _does_ make sense to fix this one, because `DiscoveryService` doesn't _necessarily_ live as long as the application

File: tracer/src/Datadog.Trace/Sampling/ManagedTraceSampler.cs (hunk 8 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Sampling/ManagedTraceSampler.cs]: Yeah, it's not worth the complexity for the rare updates here I think. Will switch to `lock` instead.

File: tracer/src/Datadog.Trace/Agent/Api.cs (hunk 2 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/Api.cs]: Fixed

**Comment by andrewlock** [comment]
# Code Review: PR #7724 - Handle changes to MutableSettings and ExporterSettings without rebuilding

## Overview

This PR implements dynamic reconfiguration of tracer services in response to runtime settings changes (from manual configuration or remote config), eliminating the need to tear down and rebuild the entire tracer infrastructure. The changes introduce "managed" versions of services that can respond to specific setting changes without full disposal/reconstruction.

**Key Changes:**
- Introduce `IStatsdManager` abstraction with lease-based access pattern
- Add `ManagedApi` and update `DiscoveryService` to handle exporter setting changes
- Update `AgentWriter`, `Api`, `StatsAggregator`, and related services to subscribe to relevant setting changes
- Replace `IDogStatsd` with `IStatsdManager` throughout
- Add thread-safe patterns using `Volatile.Read()` for accessing mutable state

---

## Code Quality & Architecture

### Strengths ✓

1. **Clear separation of concerns**: Managed vs unmanaged services (e.g., `DiscoveryService.CreateManaged()` vs `CreateUnmanaged()`)
2. **Selective subscription pattern**: Services only subscribe to changes they care about
3. **Backward compatibility**: CI Vis and Debugger paths maintain existing behavior
4. **Thread safety**: Consistent use of `Volatile.Read()` for cross-thread visibility
5. **Lease pattern**: The `IStatsdManager.TryGetClientLease()` pattern is clean and prevents access to disposed/stale clients

### Areas for Improvement ⚠️

1. **Thread safety complexity**: Multiple patterns for handling concurrency could be consolidated
2. **Memory management**: No explicit disposal of old `Api` instances in `ManagedApi`
3. **Subscription lifecycle**: Some subscriptions (e.g., in `ManagedApi`) are never disposed
4. **Pattern inconsistency**: Different services use different update mechanisms

---

## Specific Issues & Suggestions

### 1. Memory Leak Risk in ManagedApi.cs:52

```csharp
var api = new Api(apiRequestFactory, statsd, updateSampleRates, partialFlushEnabled, healthMetricsEnabled);
Interlocked.Exchange(ref _api!, api);
```

**Issue**: Old `Api` instances are replaced but never explicitly disposed. The `Api` class doesn't implement `IDisposable`, but it holds resources like `IApiRequestFactory`.

**Suggestion**: Consider implementing disposal or documenting why it's safe to replace without cleanup.

### 2. Race Condition Pattern in AgentWriter.cs:138-161

```csharp
if (Volatile.Read(ref _traceMetricsEnabled))
{
    using var lease = _statsd.TryGetClientLease();
    if (lease.Client is { } statsd)
    {
        statsd.Increment(TracerMetricNames.Queue.EnqueuedTraces);
        statsd.Increment(TracerMetricNames.Queue.EnqueuedSpans, trace.Count);
    }
}
```

**Issue**: `_traceMetricsEnabled` could change between the check and the lease acquisition. While benign here, this pattern appears in multiple locations.

**Suggestion**: Consider caching the `Volatile.Read()` result to a local variable for consistency:
```csharp
var metricsEnabled = Volatile.Read(ref _traceMetricsEnabled);
if (metricsEnabled)
{
    using var lease = _statsd.TryGetClientLease();
    // ...
}
```

### 3. Inconsistent Pattern in Api.cs:314-325

In `Api.SendTracesAsyncImpl`, the lease is acquired conditionally:
```csharp
using var lease = Volatile.Read(ref _healthMetricsEnabled) ? _statsd.TryGetClientLease() : default;
var healthStats = _healthMetricsEnabled ? lease.Client : null;
```

**Issue**: Uses `_healthMetricsEnabled` without `Volatile.Read()` in the second line, but with it in the first. Also reads the field twice.

**Suggestion**:
```csharp
var healthMetricsEnabled = Volatile.Read(ref _healthMetricsEnabled);
using var lease = healthMetricsEnabled ? _statsd.TryGetClientLease() : default;
var healthStats = lease.Client;
```

### 4. ClientStatsPayload Record Pattern (ClientStatsPayload.cs:14-25)

```csharp
private AppSettings _settings = CreateSettings(settings);
// ...
public void UpdateDetails(MutableSettings settings)
    => Interlocked.Exchange(ref _settings, CreateSettings(settings));
```

**Issue**: While functionally correct, using `Interlocked.Exchange` on a record field is unconventional. Records are typically immutable by design.

**Suggestion**: Document why this pattern was chosen, or consider using a class instead of a record.

### 5. Missing Null Check in AgentWriter.cs:73-79

```csharp
settings.Manager.SubscribeToChanges(changes =>
{
    if (changes.UpdatedMutable is { } mutable
     && mutable.TracerMetricsEnabled != changes.PreviousMutable.TracerMetricsEnabled)
    {
        // ...
    }
});
```

**Issue**: If `changes.PreviousMutable` is null, accessing `TracerMetricsEnabled` could throw (though unlikely in practice).

**Suggestion**: Add null-safety check or document that `PreviousMutable` is never null.

---

## Testing Considerations

### Missing Test Coverage

1. **Concurrency tests**: No evidence of tests for race conditions during setting updates
2. **Subscription cleanup**: Tests for memory leaks from subscriptions
3. **Service reconfiguration**: Integration tests validating services continue working after settings change
4. **Lease pattern edge cases**: What happens if settings change mid-lease?

### Recommended Test Scenarios

1. Update `ExporterSettings` while traces are being sent
2. Toggle `TracerMetricsEnabled` rapidly to test race conditions
3. Update multiple settings simultaneously
4. Verify old service instances are eventually GC'd

---

## Performance Implications

### Positive Impacts ✓
- Eliminates overhead of tearing down/rebuilding entire infrastructure
- Reduces GC pressure by reusing services

### Potential Concerns ⚠️
- `Volatile.Read()` has memory barrier overhead (minimal on modern CPUs)
- Multiple subscription callbacks on hot paths (e.g., every trace write)
- Lease pattern adds indirection (negligible but measurable)

**Recommendation**: Add performance benchmarks comparing before/after settings changes.

---

## Security Considerations

No major security concerns identified. The changes are primarily architectural and don't introduce new external inputs or trust boundaries.

---

## Overall Assessment

**Rating: Approve with Minor Changes**

This is a solid architectural improvement that significantly reduces the overhead of runtime configuration changes. The implementation is generally well-thought-out with good separation between managed and unmanaged paths.

### Must-Fix Before Merge
1. Document memory management strategy for replaced `Api` instances
2. Fix inconsistent use of `Volatile.Read()` in `Api.cs:314-325`
3. Add null-safety checks in subscription lambdas

### Should-Fix Soon After
1. Add concurrency tests
2. Consolidate thread-safety patterns
3. Add XML documentation to new public APIs
4. Performance benchmarks

### Nice-to-Have
1. Consider implementing `IDisposable` on `Api` if resources need explicit cleanup
2. Add `[DebuggerDisplay]` attributes to aid debugging
3. Consistent nullable reference context across all files

The PR successfully achieves its stated goal of eliminating unnecessary service rebuilds while maintaining backward compatibility. The commit-by-commit structure mentioned in the description should make review easier.

---

*Review generated by Claude Code*

File: tracer/src/Datadog.Trace/Agent/Api.cs [hunk 9]
```diff
@@ -298,10 +311,12 @@ private async Task<SendResult> SendTracesAsyncImpl(IApiRequest request, bool fin
 
             try
             {
+                using var lease = Volatile.Read(ref _healthMetricsEnabled) ? _statsd.TryGetClientLease() : default;
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/Api.cs]: **Thread Safety Concern: Missing volatile read before lease acquisition**

In this method, you're checking `_healthMetricsEnabled` without a volatile read before getting the client lease, but then using `Volatile.Read()` inside the ternary. This creates a potential race condition where the health metrics setting could be disabled between the two checks, leading to unnecessary lease acquisition.

Suggested fix:
```csharp
var healthMetricsEnabled = Volatile.Read(ref _healthMetricsEnabled);
using var lease = healthMetricsEnabled ? _statsd.TryGetClientLease() : default;
var healthStats = healthMetricsEnabled ? lease.Client : null;
```

This ensures consistent state throughout the method and avoids the double-check pattern that could cause issues.

🤖 Claude Code

File: tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs [hunk 10]
```diff
@@ -32,15 +32,35 @@ internal class DiscoveryService : IDiscoveryService
...
+            int recheckIntervalMs)
+            : this(CreateApiRequestFactory(settings.InitialExporterSettings, tcpTimeout), initialRetryDelayMs, maxRetryDelayMs, recheckIntervalMs)
+        {
+            // Create as a "managed" service that can update the request factory
+            _settingSubscription = settings.SubscribeToChanges(changes =>
+            {
+                if (changes.UpdatedExporter is { } exporter)
+                {
+                    var newFactory = CreateApiRequestFactory(exporter, tcpTimeout);
+                    Interlocked.Exchange(ref _apiRequestFactory!, newFactory);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs:63]: **Potential NullReferenceException: Missing null check after Exchange**

After `Interlocked.Exchange(ref _apiRequestFactory!, newFactory)`, the old factory is not disposed. While this may be intentional for thread safety reasons, there's no guarantee that `_apiRequestFactory` is initialized before the subscription callback fires.

The constructor calls `CreateApiRequestFactory()` after setting up the subscription, which means the callback could theoretically fire before `_apiRequestFactory` is initialized (though unlikely in practice).

Consider either:
1. Initialize `_apiRequestFactory` before creating the subscription
2. Add a null-check in the subscription callback
3. Dispose the old factory if it's safe to do so

🤖 Claude Code

File: tracer/src/Datadog.Trace/Agent/AgentWriter.cs [hunk 11]
```diff
@@ -138,10 +151,14 @@ public void WriteTrace(ArraySegment<Span> trace)
                 }
             }
 
-            if (_statsd != null)
+            if (Volatile.Read(ref _traceMetricsEnabled))
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/AgentWriter.cs:154]: **Thread Safety Issue: TOCTOU (Time-of-Check-Time-of-Use) vulnerability**

This pattern has a classic TOCTOU race condition:

1. Line 154: Check `Volatile.Read(ref _traceMetricsEnabled)`
2. Line 156: Get client lease
3. Line 157: Null-check the client

Between steps 1 and 2, another thread could disable metrics and dispose the statsd client, potentially causing issues when `TryGetClientLease()` is called on a disposed/null client.

Suggested fix - cache the volatile read:
```csharp
var metricsEnabled = Volatile.Read(ref _traceMetricsEnabled);
if (metricsEnabled)
{
    using var lease = _statsd.TryGetClientLease();
    if (lease.Client is { } statsd)
    {
        statsd.Increment(TracerMetricNames.Queue.EnqueuedTraces);
        statsd.Increment(TracerMetricNames.Queue.EnqueuedSpans, trace.Count);
    }
}
```

This pattern appears multiple times in AgentWriter.cs (lines 154, 334, 536) and should be fixed consistently.

🤖 Claude Code

File: tracer/src/Datadog.Trace/Agent/ManagedApi.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+                _api.ToggleTracerHealthMetrics(mutable.TracerMetricsEnabled);
+            }
+        });
+
+        [MemberNotNull(nameof(_api))]
+        void UpdateApi(ExporterSettings exporterSettings, bool healthMetricsEnabled)
+        {
+            var apiRequestFactory = TracesTransportStrategy.Get(exporterSettings);
+            var api = new Api(apiRequestFactory, statsd, updateSampleRates, partialFlushEnabled, healthMetricsEnabled);
+            Interlocked.Exchange(ref _api!, api);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/ManagedApi.cs:52]: **Potential Resource Leak: No disposal of old `_api` instances**

In `UpdateApi()`, you're using `Interlocked.Exchange(ref _api!, api)` to swap the API instance, but the old instance is never disposed. Over time, if settings change frequently, this could lead to resource leaks (open connections, tasks, etc.).

The `Api` class doesn't implement `IDisposable`, so this might be by design, but it's worth confirming that the old instances are properly garbage collected and don't hold onto resources.

Consider either:
1. Making `Api` disposable and disposing the old instance
2. Adding a comment explaining why disposal isn't necessary
3. Implementing a proper cleanup mechanism

🤖 Claude Code

File: tracer/src/Datadog.Trace/Agent/AgentWriter.cs [hunk 13]
```diff
@@ -65,17 +65,28 @@ internal class AgentWriter : IAgentWriter
 
         private long _droppedTraces;
 
-        public AgentWriter(IApi api, IStatsAggregator statsAggregator, IDogStatsd statsd, TracerSettings settings)
-            : this(api, statsAggregator, statsd, maxBufferSize: settings.TraceBufferSize, batchInterval: settings.TraceBatchInterval, apmTracingEnabled: settings.ApmTracingEnabled)
+        private bool _traceMetricsEnabled;
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/AgentWriter.cs:68]: **Missing Initialization: `_traceMetricsEnabled` used before assignment**

The field `_traceMetricsEnabled` is declared but not initialized. The first constructor (starting line 70) calls the internal constructor and sets up a subscription that immediately references `_traceMetricsEnabled` via `Volatile.Write()`.

While the subscription callback won't fire until after construction completes, it's better to explicitly initialize the field to avoid potential issues and make the intent clear.

Add initialization:
```csharp
private bool _traceMetricsEnabled = false;
```

🤖 Claude Code

**Comment by andrewlock** [comment]
## Review Summary

This is a significant refactoring that transitions from rebuilding all services when settings change to dynamically updating only the necessary components. The overall architecture is sound and the commit-by-commit approach makes the changes manageable.

### Key Findings

**Critical Issues:**
1. **Thread Safety Concerns**: Multiple TOCTOU (Time-of-Check-Time-of-Use) race conditions in  where  checks are not cached before use, potentially causing issues if settings change mid-execution.

2. **Resource Management**: The new managed services (, ) swap underlying instances without disposing old ones. While this may be intentional for thread safety, it risks resource leaks if settings change frequently.

**Moderate Issues:**
3. **Field Initialization**: The  field in  lacks explicit initialization, relying on default bool values.

4. **Inconsistent Volatile Access**: In , the  field changed from readonly to mutable, but initial access in  doesn't use volatile reads consistently.

The patterns introduced here (subscription-based updates, volatile reads, managed wrappers) are appropriate for the goal, but need tightening around thread safety guarantees. I recommend reviewing all  usage to ensure values are cached locally before use, and documenting the lifecycle expectations for swapped service instances.

Review by 🤖 Claude Code

File: tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs (hunk 10 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs:63]: I think this is nonsense:
1. the factory doesn't need to be disposed (it doesn't implement IDisposable)
2. This is only called _after_ the other constructor, so _apiRequestFactory is not null when this is called

File: tracer/src/Datadog.Trace/Agent/AgentWriter.cs (hunk 11 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/AgentWriter.cs:154]: I think this is fine 🤔 The `metricsEnabled` check is just to avoid calling `TryGetClientLease()` when we don't need it. The StatsDManager won't give out an already disposed lease, so there's nothing to worry about threading-wise I believe. 

Also the suggested code is functionally identical

File: tracer/src/Datadog.Trace/Agent/ManagedApi.cs (hunk 12 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/ManagedApi.cs:52]: There's nothing disposable, none of this stuff is disposable

File: tracer/src/Datadog.Trace/Agent/AgentWriter.cs (hunk 13 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/AgentWriter.cs:68]: That's automatic in C#

File: tracer/src/Datadog.Trace/Agent/ManagedApi.cs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+                UpdateApi(exporter, mutable.TracerMetricsEnabled);
+            }
+            else if (changes.UpdatedMutable is { } mutable && mutable.TracerMetricsEnabled != changes.PreviousMutable.TracerMetricsEnabled)
+            {
+                _api.ToggleTracerHealthMetrics(mutable.TracerMetricsEnabled);
+            }
+        });
+
+        [MemberNotNull(nameof(_api))]
+        void UpdateApi(ExporterSettings exporterSettings, bool healthMetricsEnabled)
```
Comment (lucaspimentel) [inline tracer/src/Datadog.Trace/Agent/ManagedApi.cs:48]:
```suggestion
        static void UpdateApi(ExporterSettings exporterSettings, bool healthMetricsEnabled)
```

Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/ManagedApi.cs:48]: This can't be static, it mutates `_api` (as indicated by [MemberNotNull] 🙂)

File: tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs [hunk 15]
```diff
@@ -200,7 +217,9 @@ public void InjectPathwayContextAsBase64String<TCarrier>(PathwayContext? context
             var edgeStartNs = previousContext == null && timeInQueueMs > 0 ? nowNs - (timeInQueueMs * 1_000_000) : nowNs;
             var pathwayStartNs = previousContext?.PathwayStart ?? edgeStartNs;
 
-            var nodeHash = HashHelper.CalculateNodeHash(_nodeHashBase, edgeTags);
+            // Don't blame me, blame the fact we can't do Volatile.Read with a ulong in .NET FX...
+            var nodeHashBase = new NodeHashBase(unchecked((ulong)Volatile.Read(ref Unsafe.As<ulong, long>(ref _nodeHashBase))));
```
Comment (bouwkast) [inline tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs]: 😅 

So uh what exactly is going on here?
It is a `struct` so can't use InterLocked.CompareExchange but we also need to convert it from ulong to long?

File: tracer/src/Datadog.Trace/Agent/ManagedApi.cs (hunk 14 above)
Comment (lucaspimentel) [reply tracer/src/Datadog.Trace/Agent/ManagedApi.cs:48]: Weird, before commenting, I actually checked out the code and added `static`. I guess I was just looking for the IDE's red squiggles and didn't actually _build_ the code 😅

Comment (lucaspimentel) [reply tracer/src/Datadog.Trace/Agent/ManagedApi.cs:48]: Nope, yep. I see them squiggles now.

File: tracer/test/snapshots/OpenTelemetrySdkTests.SubmitsOtlpMetrics_up_to_1_5_0.NET_6_DD.verified.txt [hunk 16]
```diff
@@ -25,7 +25,7 @@
             {
               "key": "service.name",
               "value": {
-                "string_value": "unknown_service:dotnet"
+                "string_value": "Samples.OpenTelemetrySdk"
```
Comment (bouwkast) [inline tracer/test/snapshots/OpenTelemetrySdkTests.SubmitsOtlpMetrics_up_to_1_5_0.NET_6_DD.verified.txt:28]: Hmm I do remember running into this previously with `unknown_service` I think I did something to override this like so a long time ago, unsure how / why this changed though now 🤷

Comment (andrewlock) [reply tracer/test/snapshots/OpenTelemetrySdkTests.SubmitsOtlpMetrics_up_to_1_5_0.NET_6_DD.verified.txt:28]: Yeah, I'm going to defer to @link04 on this one 😅 My guess is that there were/are some race conditions here. I think this is now "correct" to be fair 😄

File: tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs (hunk 15 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs]: _so_, yeah, this is a mess 😅 

- We want to cache the value of `NodeHashBase`
- `NodeHashBase` depends on some mutable settings, e.g. service name, so we need to be able to _replace_ `NodeHashBase`.
- We need to use `Interlocked.Exchange()` to make sure we do the swap atomically.
  - You _can't_ do `Interlocked.Exchange()` with a struct (for technical reasons) 🙁 
- Luckily, `NodeHashBase` is just a thin wrapper around a `ulong`, which is 64 bit and _can_ be used with  `Interlocked.Exchange` 🎉 
  - _Except_, < .NET 5, you _can't_ do `Interlocked.Exchange(ulong)` because the overloads don't exist 😭 
- The "solution" for _updating_ is (you can see this in the `UpdateNodeHash` method):
    - Calculate the new `NodeHashBase`
    - Grab the `ulong` from the `NodeHashBase`
    - Do an unchecked cast of the `ulong` as a `long`
    - Do an `Interlocked.Exchange(ref long)`, treating the stored `ulong _nodeHashBase` as a `long` for the purposes of the exchange
- The solution for _reading_ is:
  - Do a `Volatile.Read(ref long)`, treating the stored `ulong _nodeHashBase` as a `long` for the purposes of the read
    - Do an unchecked case of the `long` as a `ulong`,
    - Create the new `NodeHashBase`

Yes, it's a PITA. 

All that said, I just realised that if we store the `_nodeHashBase` as an `int`, we can avoid all the `Unsafe.As` reinterpretation, and just stick with the unchecked casts, which simplifies this a bit! 🙂

**Review by tonyredondo — APPROVED** [review]
